java-checks/src/main/java/org/sonar/java/checks/security/PasswordEncoderCheck.java

```diff
@@ -87,6 +129,92 @@ public void visitNode(Tree tree) {
+      if (!Objects.equals(iterationCountValueExpression.firstToken(), iterationCountExpression.firstToken())) {
+        secondaryLocations.add(new Location("", iterationCountValueExpression));
+      }
+
+      reportIssue(iterationCountExpression, JAVAX_CRYPTO_MESSAGE_FORMAT.formatted(minIteration), secondaryLocations, null);
+    }
+  }
+
+  // Given secretKeyFactory.generateSecret(keySpec), where var keySpec = new PBEKeySpec(..., iterations, ...), where
+  // iterations is var iterations = y, returns the expressions "iterations", "y", as well as "y" value (e.g. 120000).
+  private static Optional<ExpressionsAndValue<Integer>> extractIterationCount(ExpressionTree keySpecArgumentExpression) {
+    return getInitializerExpressionOfVariableIdentifier(keySpecArgumentExpression)
+      // try to resolve the expression itself to a constructor invocation
+      .or(() -> Optional.of(keySpecArgumentExpression))
+      .filter(e -> e.is(Tree.Kind.NEW_CLASS))
+      .map(NewClassTree.class::cast)
+      .filter(PBE_KEY_SPEC_CONSTRUCTOR::matches)
+      .map(gi -> gi.arguments().get(2))
+      .flatMap(e -> getInitializerExpressionAndValue(e, Integer.class));
+  }
+
+  // Given secretKeyFactory.getInstance(algorithm), where var algorithm = y, returns the expressions "algorithm",
+  // "y", as well as "y" value (e.g. "PBKDF2withHmacSHA512").
+  private static Optional<ExpressionsAndValue<String>> extractAlgorithm(ExpressionTree secretKeyFactoryExpression) {
+    return getInitializerExpressionOfVariableIdentifier(secretKeyFactoryExpression)
+      // try to resolve the expression itself to a method invocation
+      .or(() -> Optional.of(secretKeyFactoryExpression))
+      .filter(e -> e.is(Tree.Kind.METHOD_INVOCATION))
+      .map(MethodInvocationTree.class::cast)
+      .filter(SECRET_KEY_FACTORY_GET_INSTANCE_METHOD::matches)
+      .map(gi -> gi.arguments().get(0))
+      .flatMap(e -> getInitializerExpressionAndValue(e, String.class));
+  }
+
+  // Given any expression e, attempts constant value resolution of the expression itself and of the initializer
+  // of the declaration of e, if any.
+  private static <T> Optional<ExpressionsAndValue<T>> getInitializerExpressionAndValue(ExpressionTree expression, Class<T> type) {
+    var initializerExpression = getInitializerExpressionOfVariableIdentifier(expression);
+    return initializerExpression
+      .flatMap(e -> e.asConstant(type))
+      .map(v -> new ExpressionsAndValue<>(expression, initializerExpression.get(), v))
+      .or(() -> expression.asConstant(type).map(t -> new ExpressionsAndValue<>(expression, expression, t)));
+  }
+
+  // Given any identifier "x", returns the initializer y from the declaration of x: "var x = y" (whether in the same
+  // statement or not, whether it was reassigned between declaration and reference or not).
+  private static Optional<ExpressionTree> getInitializerExpressionOfVariableIdentifier(ExpressionTree expression) {
```

Comment:
I think we can update a bit the name to make it clearer. The return type already gives us information about the expression and the if make clearer when we return a value.
```suggestion
  private static Optional<ExpressionTree> getInitializerIfExpressionIsVariableIdentifier(ExpressionTree expression) {
```

Reply:
Sure, it definitely reads better!